Docs: Test examples that recreate lang analyzers

nik9000 opened the pull request:
We have a pile of documentation describing how to rebuild the built in
language analyzers and, previously, our documentation testing framework
made sure that the examples successfully built *an* analyzer but they
didn't assert that the analyzer built by the documentation matches the
built in anlayzer. Unsuprisingly, some of the examples aren't quite
right.

This adds a mechanism that tests that the analyzers built by the docs.
The mechanism is fairly simple and brutal but it seems to be working:
build a hundred random unicode sequences and send them through the
`_analyze` API with the rebuilt analyzer and then again through the
built in analyzer. Then make sure both APIs return the same results.
Each of these calls to `_anlayze` takes about 20ms on my laptop which
seems fine.

Related to #29499

nik9000 committed:
Wip

nik9000 committed:
Merge branch 'master' into builtin_analyzer_tests

nik9000 committed:
Docs: Test examples that recreate lang analyzers

We have a pile of documentation describing how to rebuild the built in
language analyzers and, previously, our documentation testing framework
made sure that the examples successfully built *an* analyzer but they
didn't assert that the analyzer built by the documentation matches the
built in anlayzer. Unsuprisingly, some of the examples aren't quite
right.

This adds a mechanism that tests that the analyzers built by the docs.
The mechanism is fairly simple and brutal but it seems to be working:
build a hundred random unicode sequences and send them through the
`_analyze` API with the rebuilt analyzer and then again through the
built in analyzer. Then make sure both APIs return the same results.
Each of these calls to `_anlayze` takes about 20ms on my laptop which
seems fine.

nik9000 committed:
Fix names

elasticmachine commented:
Pinging @elastic/es-core-infra

nik9000 commented:
I've filed this as ":core/build" because ":core/build" includes testing infrastructure. It might be more correct to file it as ":search/anlaysis" but the bulk of the work is in the testing infrastructure.

I thought of three ways to do this:
1. Write a unit test that parses the docs looking for these snippets and runs the tests.
2. Create special syntax in the docs tests that kicks off these test.
3. Allow the docs tests to write things in YAML and have that kick off the test.

I opted for option 3 because it seemed the simplest thing at the time. Option two would have been fairly simple as well. Option 1 might produce faster tests but would require a lot of extra complexity that reproduces some of the docs testing infrastructure. Since option 3 takes ~20 milliseconds per language I think we're ok as far as speed is concerned.

nik9000 commented:
We also talking about doing this using some mechanism to force us to manually check the lucene analyzers when we upgrade lucene. Given how quickly these tests run I don't think that is worth it.

mayya-sharipova commented on docs/src/test/java/org/elasticsearch/smoketest/DocsClientYamlTestSuiteIT.java:
```diff
@@ -64,5 +89,89 @@ protected ClientYamlTestClient initClientYamlTestClient(ClientYamlSuiteRestSpec
+            for (int i = 0; i < size; i++) {
+                testText.add(randomRealisticUnicodeOfCodepointLength(between(1, 15))
+                    // Don't look up stashed values
+                    .replace("$", "\\$"));
```
Do strings generated from `randomRealisticUnicodeOfCodepointLength` also contain spaces, punctuation marks etc, so that we can test tokenizer part of analyzers?

nik9000 replied:
It [doesn't](https://github.com/randomizedtesting/randomizedtesting/blob/master/randomized-runner/src/main/java/com/carrotsearch/randomizedtesting/generators/RealisticUnicodeGenerator.java) look like they do. It'd be cool to insert spaces. I don't think I could easily use the *same* unicode page for both space separated strings though. That might not be too bad though.

mayya-sharipova commented on docs/reference/analysis/analyzers/lang-analyzer.asciidoc:
```diff
@@ -97,13 +97,14 @@ PUT /arabic_example
             "arabic_stop",
-            "arabic_normalization",
             "arabic_keywords",
+            "arabic_normalization",
```
+1 for correcting documentation on all analyzers

nik9000 committed:
Merge branch 'master' into builtin_analyzer_tests

nik9000 committed:
Spaces

nik9000 committed:
Document sytax enhancement

nik9000 commented:
@mayya-sharipova, I've pushed a patch to add spaces. It isn't perfect, but it does add spaces. Have a look at the comment I sent for a more thorough explanation of what I mean.

mayya-sharipova reviewed (changes requested):
@nik9000  Thanks for the change, Nik! For the search side (analyzers and a test to test that tokens are similar),  I have left a small comment.  Other than that everything looks fine.
May be somebody from the Core/Infra team can review it as well for the testing infrastructure part.

nik9000 commented on docs/src/test/java/org/elasticsearch/smoketest/DocsClientYamlTestSuiteIT.java:
```diff
@@ -64,5 +91,115 @@ protected ClientYamlTestClient initClientYamlTestClient(ClientYamlSuiteRestSpec
+                if (false == secondTokens.hasNext()) {
+                    fail(second + " has fewer tokens than " + first + ". "
+                        + first + " has [" + firstTokens.next() + "] but " + second + " is out of tokens. "
+                        + first + "'s last token was [" + previousFirst + "] and "
```
Yeah. I used to have it working. I'll fix.

nik9000 committed:
Merge branch 'master' into builtin_analyzer_tests

nik9000 committed:
Fixes

nik9000 committed:
Merge branch 'master' into builtin_analyzer_tests

nik9000 committed:
Merge branch 'master' into builtin_analyzer_tests

nik9000 committed:
Move flag

This makes the change to the regex smaller and fixes some parse errors I
hadn't noticed before.

nik9000 committed:
Merge branch 'master' into builtin_analyzer_tests

nik9000 committed:
Fix up irish stemmer

nik9000 committed:
,

nik9000 committed:
Fix irish better

nik9000 committed:
Fix cjk

nik9000 committed:
.

nik9000 committed:
,

nik9000 committed:
, againt

nik9000 committed:
Sigh

nik9000 commented:
@elasticmachine recheck this please.

nik9000 committed:
Merge branch 'master' into builtin_analyzer_tests

nik9000 committed:
Fix precommit

nik9000 committed:
Merge branch 'master' into builtin_analyzer_tests

nik9000 committed:
Remove errant class file

dakrone commented on docs/src/test/java/org/elasticsearch/smoketest/DocsClientYamlTestSuiteIT.java:
```diff
@@ -64,5 +90,117 @@ protected ClientYamlTestClient initClientYamlTestClient(ClientYamlSuiteRestSpec
+        private static CompareAnalyzers parse(XContentParser parser) throws IOException {
+            XContentLocation location = parser.getTokenLocation();
+            CompareAnalyzers section = PARSER.parse(parser, location);
+            assert parser.currentToken() == Token.END_OBJECT;
```
Can you add a message here so it'll be helpful if someone accidentally misses a closing token?

dakrone commented on test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/ExecutableSection.java:
```diff
@@ -26,15 +26,18 @@
+     * Default list of {@link ExecutableSection}s available for tests.
      */
-    NamedXContentRegistry XCONTENT_REGISTRY = new NamedXContentRegistry(Arrays.asList(
+    List<NamedXContentRegistry.Entry> DEFAULT_EXECUTABLE_CONTEXTS = unmodifiableList(Arrays.asList(
```
I think this can be final

nik9000 replied:
I started out declaring it `public static final` out of habit but checkstyle failed because they are all forced on that field because it is in an interface.

dakrone commented on test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/ExecutableSection.java:
```diff
@@ -46,6 +49,12 @@
+     * {@link NamedXContentRegistry} that parses the default list of
+     * {@link ExecutableSection}s available for tests.
+     */
+    NamedXContentRegistry XCONTENT_REGISTRY = new NamedXContentRegistry(DEFAULT_EXECUTABLE_CONTEXTS);
```
Same here, this could be final I think?

dakrone reviewed (approved):
LGTM, I left really minor nits, but nothing that needs another review

nik9000 committed:
Merge branch 'master' into builtin_analyzer_tests

nik9000 committed:
Add warning

nik9000 committed:
Merge branch 'master' into builtin_analyzer_tests
